### PR TITLE
Refine channel test

### PR DESCRIPTION
**Pull request by chengduoZH**
fix https://github.com/PaddlePaddle/Paddle/issues/7948
- [x] The basic function of `buffered_channel`.
- [x] The function of `buffered_channel` under multithreading.

### CODE REVIEW

File: paddle/framework/channel_test.cc [hunk 1]
```diff
@@ -11,16 +11,58 @@ distributed under the License is distributed on an "AS IS" BASIS,
...
+TEST(Channel, MultiThread) {
+  namespace framework = paddle::framework;
+
+  const int capacity = 10;
+  Channel<int>* ch = MakeChannel<int>(capacity);
+
+  framework::ThreadPool* pool;
+  pool = framework::ThreadPool::GetInstance();
+
+  // Receiver
```
Comment (typhoonzero) [inline paddle/framework/channel_test.cc]: Can be "Consumer" and "Producer".

File: paddle/framework/channel_test.cc [hunk 2]
```diff
@@ -11,16 +11,58 @@ distributed under the License is distributed on an "AS IS" BASIS,
...
 
 TEST(Channel, MakeAndClose) {
-  using paddle::framework::Channel;
-  using paddle::framework::MakeChannel;
-  using paddle::framework::CloseChannel;
+  Channel<int>* ch = MakeChannel<int>(10);
+  CloseChannel(ch);
+}
 
+TEST(Channel, Buffered) {
```
Comment (typhoonzero) [inline paddle/framework/channel_test.cc]: Do we need a test case for "unbuffered" channel too?

Comment (chengduoZH) [reply paddle/framework/channel_test.cc]: The implement of `unbuffered` channel is empty.
https://github.com/PaddlePaddle/Paddle/blob/fb1a0dfb157032f3eb7cdabcb445a77656ef2498/paddle/framework/details/unbuffered_channel.h#L41-L48

File: paddle/framework/channel_test.cc (hunk 1 above)
Comment (chengduoZH) [reply paddle/framework/channel_test.cc]: Done.

File: paddle/framework/channel_test.cc (hunk 2 above)
Comment (chengduoZH) [reply paddle/framework/channel_test.cc]: This [issue](https://github.com/PaddlePaddle/Paddle/issues/7937) is about `unbuffered` channel.

File: paddle/framework/channel_test.cc [hunk 3]
```diff
@@ -11,16 +11,58 @@ distributed under the License is distributed on an "AS IS" BASIS,
...
+}
+
+TEST(Channel, MultiThread) {
+  namespace framework = paddle::framework;
+
+  const int capacity = 10;
+  Channel<int>* ch = MakeChannel<int>(capacity);
+
+  framework::ThreadPool* pool;
+  pool = framework::ThreadPool::GetInstance();
```
Comment (wangkuiyi) [inline paddle/framework/channel_test.cc]: For two reasons, let us do not use `framework::ThreadPool` in this unit test, but use `std::thread`:

1. To minimize the dependencies.
1. `framework::ThreadPool` sets the number of OS threads automatically, which could be one.

File: paddle/framework/channel_test.cc [hunk 4]
```diff
@@ -11,16 +11,58 @@ distributed under the License is distributed on an "AS IS" BASIS,
...
+
+  // Consumer
+  for (int i = 0; i < capacity; ++i) {
```
Comment (wangkuiyi) [inline paddle/framework/channel_test.cc]: I don't see the purpose of this unit test. Do you have a list of edge cases?

File: paddle/framework/channel.h [hunk 5]
```diff
@@ -50,6 +51,11 @@ Channel<T>* MakeChannel(size_t buffer_size) {
 
 template <typename T>
 void CloseChannel(Channel<T>* ch) {
+  ch->Close();
+}
+
+template <typename T>
+void DeleteChannel(Channel<T>* ch) {
```
Comment (wangkuiyi) [inline paddle/framework/channel.h]: I agree that we need CloseChannel, which mimics Go's close, and DeleteChannel, which mimics Go's garbage collection.

One step forward, I think we can have only CloseChannel, because DeleteChannel could be simplified as `delete ch` -- I didn't have to and should not have defined `template <typename T> void DeleteChannel`.

File: paddle/framework/channel_test.cc [hunk 6]
```diff
@@ -72,4 +72,5 @@ TEST(Channel, ConcurrentSendNonConcurrentReceiveWithSufficientBufferSize) {
   std::this_thread::sleep_for(std::chrono::milliseconds(100));  // wait 0.5 sec
   EXPECT_EQ(sum, 45U);
   CloseChannel(ch);
+  t.join();
```
Comment (wangkuiyi) [inline paddle/framework/channel_test.cc]: It seems that you closed but didn't delete the channel in this test?  In my mind, this test should look like

```c++
std::unique_ptr<Channel<int>*> ch(MakeChannel<int>(10)); // `delete ch.get()` will be called
do_something_here_with(ch);
CloseChannel(ch.get());
```

File: paddle/framework/details/buffered_channel.h [hunk 7]
```diff
@@ -33,15 +33,17 @@ class Buffered : public paddle::framework::Channel<T> {
...
   virtual size_t Cap() { return cap_; }
+  virtual void Close();
 
  private:
   size_t cap_;
   std::mutex mu_;
   std::condition_variable empty_cond_var_;
   std::condition_variable full_cond_var_;
   std::deque<T> channel_;
+  bool close;
```
Comment (wangkuiyi) [inline paddle/framework/details/buffered_channel.h]: `close` => `closed_` or `closing_`

Comment (chengduoZH) [reply paddle/framework/details/buffered_channel.h]: Done

File: paddle/framework/channel.h (hunk 5 above)
Comment (chengduoZH) [reply paddle/framework/channel.h]: I agree with you. 
We should make a further analysis about `CloseChannel`. When the channel was closed, the `channel_`  of `Channel` may also have some data, I think that `receive` can also work properly, but `send` cannot send data to `channel_`. 
So, `DeleteChannel` should not delete `Buffered` or `Unbuffered`.

File: paddle/framework/channel_test.cc (hunk 6 above)
Comment (chengduoZH) [reply paddle/framework/channel_test.cc]: > It seems that you closed but didn't delete the channel in this test?

We can add `DeleteChannel` behind `CloseChannel`.

Comment (wangkuiyi) [reply paddle/framework/channel_test.cc]: Maye the use of unique_ptr didn't present my idea clearly. Let us use `delete` explicitly:

```c++
auto ch = MakeChannel<int>(10);
do_something(ch);
CloseChannel(ch);
delete ch;
```

Comment (chengduoZH) [reply paddle/framework/channel_test.cc]: Ok!
If we use `delete` to destroy the `ch` but other threads also hold the `ch` and these threads send data to the `ch`, there will cause an error. 
I wonder that whether this case is existent.
